### ISSUE TITLE
Macos wpk upgrade fails with timeout using the force option

### ISSUE THREAD

**Issue by rauldpm** [wazuh/wazuh#12479]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| v4.3.0-rc4 | Upgrade | Agent | Packages | Macos Catalina (agent)<br>Amazon Linux 2 (manager) |

Hello team, testing the v4.3.0-rc4 wpk, I have found that the upgrade in Macos fails when using the option -F (forcing the agent re-install)

Throughout the wait, neither the manager log nor the agent log show activity, being the last log record at the same time when the update command was executed.

In the end, after ~15 minutes wait, upgrade fails with:

```
[root@ip-172-31-88-19 ec2-user]# /var/ossec/bin/agent_control -l

Wazuh agent_control. List of available agents:
   ID: 000, Name: ip-172-31-88-19.ec2.internal (server), IP: 127.0.0.1, Active/Local
   ID: 001, Name: macos-1015, IP: any, Active

List of agentless devices:

[root@ip-172-31-88-19 ec2-user]# /var/ossec/bin/agent_upgrade -a 001 -f /home/ec2-user/wazuh_agent_v4.3.0_macos_x86_64.wpk -x upgrade.sh -F

Upgrading...

Failed upgrades:
	Agent 001 status: Timeout reached while waiting for the response from the agent
```

When removing the agent with manage_agents, errors appear in the log:

```
2022/03/01 18:27:08 wazuh-remoted: WARNING: (1408): Invalid ID 001 for the source ip: 'xxxxxx' (name 'unknown').
2022/03/01 18:28:07 wazuh-remoted: ERROR: Response timeout for request counter '26d577ad'.
2022/03/01 18:28:07 wazuh-modulesd:agent-upgrade: ERROR: (8117): Unknown error from agent.
2022/03/01 18:28:07 wazuh-modulesd:agent-upgrade: ERROR: (8119): There has been an error updating task state. Error code: '3', message: 'No task in DB'
```

Before and after this process it has been possible to observe alerts referring to the agent

Regards, Raúl.

**Comment by rauldpm** [comment]
### Update 

Running the agent with agent.debug=2 and running the upgrade by WPK, the following is observed in the agent log:

- Matches of the word upgrade in ossec.log

```
2022/03/02 08:07:03 wazuh-modulesd[3516] main.c:94 at main(): DEBUG: Created new thread for the 'agent-upgrade' module.
2022/03/02 08:07:03 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:96 at wm_agent_upgrade_start_agent_module(): INFO: (8153): Module Agent Upgrade started.
2022/03/02 08:09:31 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb8437 upgrade {"command":"open","parameters":{"mode":"wb","file":"wazuh_agent_v4.3.0_macos_x86_64.wpk"}}'
2022/03/02 08:09:31 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"open","parameters":{"mode":"wb","file":"wazuh_agent_v4.3.0_macos_x$
2022/03/02 08:09:31 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:31 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb8438 upgrade {"command":"write","parameters":{"buffer":"V1BLMjU2AC0tLS0tQkVHSU4gQ0VSVElGSUNBVEUtLS0tLQpNSUlEQ2pDQ0FmSUNDUUNQQjk2$
2022/03/02 08:09:31 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"V1BLMjU2AC0tLS0tQkVHSU4gQ0VSVElGSUN$
2022/03/02 08:09:31 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:31 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb8439 upgrade {"command":"write","parameters":{"buffer":"Z2RmSWRVTDE4TnFVVHBieXBlSU03V2JFcWF4VjYybzg0TVVJZndNUHgyClM4bUNJbXh4bDkw$
2022/03/02 08:09:31 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"Z2RmSWRVTDE4TnFVVHBieXBlSU03V2JFcWF$
2022/03/02 08:09:31 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:31 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb843a upgrade {"command":"write","parameters":{"buffer":"Rk4xb3ZLSFVuUmNvbDhJQWlnaVdTMk1udDFra3pmWi9uM05YVzNmcENNWEdnWjFIVloKaGMw$
2022/03/02 08:09:31 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"Rk4xb3ZLSFVuUmNvbDhJQWlnaVdTMk1udDF$
2022/03/02 08:09:31 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:31 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb843b upgrade {"command":"write","parameters":{"buffer":"QUvPxOBgbQ7jYehEDMADgAD+uFDb/vyV0nr/AvFYTlbXni1sw/BtHX27VM0TtxYTvGS+c5UP$
2022/03/02 08:09:31 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"QUvPxOBgbQ7jYehEDMADgAD+uFDb/vyV0nr$
2022/03/02 08:09:31 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:32 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb843c upgrade {"command":"write","parameters":{"buffer":"E9gu4xxabl/e6+hUt8cAN+dWjvMAk4KtesHF5q80SUmLYiwtMSxqzZxqsROQyA++nRvUzqIy$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"E9gu4xxabl/e6+hUt8cAN+dWjvMAk4KtesH$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:32 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb843d upgrade {"command":"write","parameters":{"buffer":"dYQtX21tEWsGqmHp4udd7jfXAqoqG8gXj11NcJOrrGFnvLpcj75KBgbLT5K0o1QayoKXEqRY$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"dYQtX21tEWsGqmHp4udd7jfXAqoqG8gXj11$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:32 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb843e upgrade {"command":"write","parameters":{"buffer":"GVUL137zNCmoOygIEK9tDGflehQoGgfWzWoJSwHscTW4Lzkz5I2I5Q18Mwh82g8UI9HVJ+M8$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"GVUL137zNCmoOygIEK9tDGflehQoGgfWzWo$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:32 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb843f upgrade {"command":"write","parameters":{"buffer":"VICVdbu7yYzKNP/iGmO42ZJxr8hNNL9DzEAAA+8vM3zJ1XmTi+uT7gjFbI7i9oDhSZx25i11$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"VICVdbu7yYzKNP/iGmO42ZJxr8hNNL9DzEA$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:32 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb8440 upgrade {"command":"write","parameters":{"buffer":"L+vueK/Bq6J3GOMyFiigXusNDwDsLqHtFbSCm6eTVFd3QoKUx4itxMk3v3kB/Rv/9wzx7e00$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"L+vueK/Bq6J3GOMyFiigXusNDwDsLqHtFbS$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:32 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb8441 upgrade {"command":"write","parameters":{"buffer":"wypMyVIQcYIEvPI0LDiYJvwXvnhDH0N1I0V9NzNwgt4sZhoKVHyJ9A2q/vC9nwRIAiU8G/Yt$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"wypMyVIQcYIEvPI0LDiYJvwXvnhDH0N1I0V$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:32 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb8442 upgrade {"command":"write","parameters":{"buffer":"xmehQBr62OVG0SoE0xDBJcq4nIvuYhZYtCYt+P7l5kagLhd2kguDpjOc0rqH9+aHa8/ykOD9$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"xmehQBr62OVG0SoE0xDBJcq4nIvuYhZYtCY$
2022/03/02 08:09:32 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:33 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb8443 upgrade {"command":"write","parameters":{"buffer":"nHR3Va2fV7V+IcgbTAtxTdZ2JgW1ywAERdov7aZmBq+KbT4MOMyYTA5qAdWFpFYHZn4ZqMYi$
2022/03/02 08:09:33 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"nHR3Va2fV7V+IcgbTAtxTdZ2JgW1ywAERdo$
2022/03/02 08:09:33 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:33 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb8444 upgrade {"command":"write","parameters":{"buffer":"EC7s4isIkdzqkBheIKVdQjECqCxcPTFS2TfR2QXX8VQrHMQIWLKKt5cvAs8FiA1aPV/INWXt$
2022/03/02 08:09:33 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"EC7s4isIkdzqkBheIKVdQjECqCxcPTFS2Tf$
2022/03/02 08:09:33 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:33 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb8445 upgrade {"command":"write","parameters":{"buffer":"oMtr9LcCJCLWE4m96oVGAIIhMwWlY9G1zpqUAKLdchu1N1UuoA8urJBXRxpg1UylKsl2VlwP$
2022/03/02 08:09:33 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"oMtr9LcCJCLWE4m96oVGAIIhMwWlY9G1zpq$
2022/03/02 08:09:33 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:33 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb8446 upgrade {"command":"write","parameters":{"buffer":"Kf5l8QP07ptaOjX6QXJOO1sRzHhT79a2we3oKbWHqKWlgbV+VOJLLuk+91eSRiZ5vI7rFlZN$
2022/03/02 08:09:33 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"Kf5l8QP07ptaOjX6QXJOO1sRzHhT79a2we3$
2022/03/02 08:09:33 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:09:33 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb8447 upgrade {"command":"write","parameters":{"buffer":"CcVE/XbT7iiRWolInOydlocUXAf5yON1iwBHeewxElzxjTzHCQEMKZkGtx4TtwgwyvyFT/JL$

```

During the upgrade process, we can see in the ossec.log messages that seem to be in loop:

```
2022/03/02 08:11:24 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb87b4 upgrade {"command":"write","parameters":{"buffer":"VTk2PNgZUMSnhN1P4MAcwfQjS37baWIb1UXBHIwgM7gvrWc1HZGZ+9PjBWIKgbsL4vRC0GddmhMelw/Yu+maEKJioDJaPV9ABrhm/tVWCkWWz0Ewz58Z/XIJET1L+w8eOjJZ45ZN38ZKzthgWRcq1FeKN2+9heSL9NL7yHA0sXYHnvf7qznBfE4ac8c/adfEV1ZVq1+ac6zr+np7HjdSIm/1U3dD4HLpnopXRYvJZAaVxdkKPW9b7mkunrH9Z5LnoLva9mXP4jEulEwft6fG+FbTDJ6c37u815uZZa86p1F36ojq8uCIkgPvmSPejslLOaZ1Vg9XIJ5ZrcXkfXrlSKkN7phcG6ICSAYV7eiw2erpYaC8klNWI5bqOL9JoJ9D2XZevBgy9MzsZlMcd2y3lctTWgNUkk6u9N6wlAnPSmwMPN4eCuKSgLLFZ1PAsVmvLChyFX2ie7RqGspveOSdxiacIg+vpoLBd2l4e+wdgjFcfoJoCPss4pSmGSWtTCD6U+pvyNGOmYgyLB9YXMRrDFgKezMNf3iQ/fXHlg6TPCMRXVBLTB/Gk11g1xkASyhQH0vahSEmbxdZbD5oQarurTVFOjEaODMJB5sSJaIf84CyWekSlS17p9jDjy2f3k71lIXG2EAKwrg7H+cNLK4+zMqrsOoRyEwhAks/Pfy9uIw=","length":512,"file":"wazuh_agent_v4.3.0_macos_x86_64.wpk"}}'
2022/03/02 08:11:24 wazuh-agentd[3469] request.c:281 at req_receiver(): DEBUG: req_receiver(): sending '{"command":"write","parameters":{"buffer":"VTk2PNgZUMSnhN1P4MAcwfQjS37baWIb1UXBHIwgM7gvrWc1HZGZ+9PjBWIKgbsL4vRC0GddmhMelw/Yu+maEKJioDJaPV9ABrhm/tVWCkWWz0Ewz58Z/XIJET1L+w8eOjJZ45ZN38ZKzthgWRcq1FeKN2+9heSL9NL7yHA0sXYHnvf7qznBfE4ac8c/adfEV1ZVq1+ac6zr+np7HjdSIm/1U3dD4HLpnopXRYvJZAaVxdkKPW9b7mkunrH9Z5LnoLva9mXP4jEulEwft6fG+FbTDJ6c37u815uZZa86p1F36ojq8uCIkgPvmSPejslLOaZ1Vg9XIJ5ZrcXkfXrlSKkN7phcG6ICSAYV7eiw2erpYaC8klNWI5bqOL9JoJ9D2XZevBgy9MzsZlMcd2y3lctTWgNUkk6u9N6wlAnPSmwMPN4eCuKSgLLFZ1PAsVmvLChyFX2ie7RqGspveOSdxiacIg+vpoLBd2l4e+wdgjFcfoJoCPss4pSmGSWtTCD6U+pvyNGOmYgyLB9YXMRrDFgKezMNf3iQ/fXHlg6TPCMRXVBLTB/Gk11g1xkASyhQH0vahSEmbxdZbD5oQarurTVFOjEaODMJB5sSJaIf84CyWekSlS17p9jDjy2f3k71lIXG2EAKwrg7H+cNLK4+zMqrsOoRyEwhAks/Pfy9uIw=","length":512,"file":"wazuh_agent_v4.3.0_macos_x86_64.wpk"}}' to socket
2022/03/02 08:11:24 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"VTk2PNgZUMSnhN1P4MAcwfQjS37baWIb1UXBHIwgM7gvrWc1HZGZ+9PjBWIKgbsL4vRC0GddmhMelw/Yu+maEKJioDJaPV9ABrhm/tVWCkWWz0Ewz58Z/XIJET1L+w8eOjJZ45ZN38ZKzthgWRcq1FeKN2+9heSL9NL7yHA0sXYHnvf7qznBfE4ac8c/adfEV1ZVq1+ac6zr+np7HjdSIm/1U3dD4HLpnopXRYvJZAaVxdkKPW9b7mkunrH9Z5LnoLva9mXP4jEulEwft6fG+FbTDJ6c37u815uZZa86p1F36ojq8uCIkgPvmSPejslLOaZ1Vg9XIJ5ZrcXkfXrlSKkN7phcG6ICSAYV7eiw2erpYaC8klNWI5bqOL9JoJ9D2XZevBgy9MzsZlMcd2y3lctTWgNUkk6u9N6wlAnPSmwMPN4eCuKSgLLFZ1PAsVmvLChyFX2ie7RqGspveOSdxiacIg+vpoLBd2l4e+wdgjFcfoJoCPss4pSmGSWtTCD6U+pvyNGOmYgyLB9YXMRrDFgKezMNf3iQ/fXHlg6TPCMRXVBLTB/Gk11g1xkASyhQH0vahSEmbxdZbD5oQarurTVFOjEaODMJB5sSJaIf84CyWekSlS17p9jDjy2f3k71lIXG2EAKwrg7H+cNLK4+zMqrsOoRyEwhAks/Pfy9uIw=","length":512,"file":"wazuh_agent_v4.3.0_macos_x86_64.wpk"}}'
2022/03/02 08:11:24 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:11:24 wazuh-agentd[3469] request.c:334 at req_receiver(): DEBUG: req_receiver(): sending '#!-req f8bb87b4 {"error":0,"message":"ok","data":[]}' to server
```

Each time, another type of interleaved message is observed:

```
2022/03/02 08:14:02 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:14:02 wazuh-agentd[3469] request.c:334 at req_receiver(): DEBUG: req_receiver(): sending '#!-req f8bb8c8d {"error":0,"message":"ok","data":[]}' to server
2022/03/02 08:14:02 wazuh-agentd[3469] notify.c:204 at run_notify(): DEBUG: Sending keep alive: #!-Darwin |macos-1015 |19.0.0 |Darwin Kernel Version 19.0.0: Thu Oct 17 16:17:15 PDT 2019; root:xnu-6153.41.3~29/RELEASE_X86_64 |x86_64 [Mac OS X|darwin: 10.15.1 (Catalina)] - Wazuh v4.3.0 / ab73af41699f13fdd81903b5f23d8d00
4a8724b20dee0124ff9656783c490c4e merged.mg
#"_agent_ip":10.0.2.15
```

When stopping the upgrade process by control-c in the manager, ossec.log continues to register this type of non-stop messages, until a `wazuh-control stop` of the agent is performed, at that time, these messages are observed:

- Agent

```
2022/03/02 08:23:41 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:165 at wm_agent_upgrade_listen_messages(): DEBUG: (8155): Incomming message: '{"command":"write","parameters":{"buffer":"J6WFuaqiM9fnWmiMebdq3oEWTOrhWHlPj3nPiHnPjHkfGvOeHfMuRr+3G9929GCNpe+Y/TaGPuX8jJj3zJj3oTHv2THvouYdaCqJNnXlNro9vWTE0Au3TxDlPFPTRoM7TZQXZ7Db7sbDYp7JIc3PUMRbBhJvtaLzyyaUEohSnkkNBhSp/6KFJqkxWj5WWCN24nOe3pu/xS8X4peKDvCRGbMeaN+pvnnt6pun1HfRwnnYvpZ/VOTxi6ySD690nXhO3mJBQ5G/uMeg17tELMFnbdLTpSo6Yqi10UBKtjJRHvt1txyasr3hiFLKnpiJVltj18JHqYYdRRsub5X7YqBv2xujUFd7LiLMV1Ra2Qdq+CwdN6qG3XqaghY4lNM5YprRKs/rAgsknsiR26ezvKyjc36IOxtCHa2+swfjEu9pDbkfCqDYHlaL/jeqqwW6/Kl8iFaLW7EZ14NtxKudg82vn0R1HGPXGIFvPRfgC00azgHDFFb7IZqpjwzcMExnboDjUOmX5NSpirWVtoUCGIAJDizQsCsbtprNNRRMjj11gboEkN7TczdoQHmVQHkQ/g3gvZa5weJpC7lGWTEeiFLBkxegpYXf6d1E+kdTtBW8RhWgdDJ4lc28SfbiqJEJLjqOtaL/Cx/izy7tCa1AtH3OlSQJmFRtXYuoLhT/sukP4JU=","length":512,"file":"wazuh_agent_v4.3.0_macos_x86_64.wpk"}}'
2022/03/02 08:23:41 wazuh-modulesd:agent-upgrade[3516] wm_agent_upgrade_agent.c:169 at wm_agent_upgrade_listen_messages(): DEBUG: (8156): Response message: '{"error":0,"message":"ok","data":[]}'
2022/03/02 08:23:41 wazuh-agentd[3469] request.c:334 at req_receiver(): DEBUG: req_receiver(): sending '#!-req f8bb9e43 {"error":0,"message":"ok","data":[]}' to server
2022/03/02 08:23:41 wazuh-modulesd:syscollector[3516] wm_syscollector.c:203 at wm_sys_stop(): INFO: Stop received for Syscollector.
2022/03/02 08:23:41 wazuh-modulesd:syscollector[3516] wm_syscollector.c:191 at wm_sys_main(): INFO: Module finished.
2022/03/02 08:23:41 wazuh-agentd[3469] receiver.c:92 at receive_msg(): DEBUG: Received message: '#!-req f8bb9e44 upgrade {"command":"write","parameters":{"buffer":"KAz3f1Ijx5ZuxGAsihbhdL8dRTi50eshxi92qgp05gZA7hKpqE5ybwUSWCNE6KKOLZxBdPGqQHSxd2aELi57jehiCgW2K6qLoovqnZwuMtFdbIosGi1SlwHIZiXLYheLJFD3cERxLwU+LJlgjgQfjidvmuRAt7/Sdou02yL9BfPbM7YzwKQjGSMq+rI5N6PdU1JT1xxqvBBA3HYd4A5B4sYJnY/zCF7Qss0i9bVJcw02aZbRJhV3wZtuuzTRhP2HCnDhnwM9B56Oq9K6r4DqSq7Dq2wdFwh3PUFke+916vAgmSiyUbRNZOvuV/SOjGexdxRlQdaLcs9BCe5k2WIYNMHQgf+fSdyY1yKpznu9TZzFdnUWAZqMZhRSfoWbWGfy39yC7xtR4ttmlX6zoiJkY4wWJKpLtIdn7Jq0HJxByM8xO45+bdLL8EnHd6n2+p13TwKsNOpo/zsWu/8hAk71XkjCceBdDdcDyk68PEonG/xSCr+NgdHoJvq6QMq5+zDFsdkgdQ3zQymDkgEz+YZBEw3x9JPRG6vn93l82XsvRV32GM79onKu7gY9Z5N2HyOZCLpIY9WXhnf8m0S5qJxN2Kjn3n7Yqiyycr0TNwsgzlwS3b8e5gZIa458qUE/5GvVocdjXiUdi0gJWVWCu0ebgdNGP8Tra3PaQsGFkCVfm2U=","length":512,"file":"wazuh_agent_v4.3.0_macos_x86_64.wpk"}}'
2022/03/02 08:23:41 wazuh-agentd[3469] request.c:157 at req_push(): ERROR: At req_push(): Target 'upgrade' refused connection. The component might be disabled
2022/03/02 08:23:41 wazuh-logcollector: INFO: (1225): SIGNAL [(15)-(Terminated: 15)] Received. Exit Cleaning...
2022/03/02 08:23:42 wazuh-syscheckd: INFO: (1225): SIGNAL [(15)-(Terminated: 15)] Received. Exit Cleaning...
2022/03/02 08:23:42 wazuh-agentd[3469] sig_op.c:49 at HandleSIG(): INFO: (1225): SIGNAL [(15)-(Terminated: 15)] Received. Exit Cleaning...
2022/03/02 08:23:42 wazuh-execd: INFO: (1314): Shutdown received. Deleting responses.
2022/03/02 08:23:42 wazuh-execd: INFO: (1225): SIGNAL [(15)-(Terminated: 15)] Received. Exit Cleaning...
```

- Manager (notes the time jump since the Upgrade command was run until the agent stopped)
```
2022/03/02 16:08:07 sca: INFO: Security Configuration Assessment scan finished. Duration: 10 seconds.
2022/03/02 16:08:19 rootcheck: INFO: Ending rootcheck scan.
2022/03/02 16:23:44 wazuh-modulesd:agent-upgrade: ERROR: (8117): Unknown error from agent.
```

**Comment by vikman90** [comment]
## Rationale

According to the logs, the manager is sending about 7 messages of 512 bytes each, every second. This is 3582 Bps. Since the WPK file is 5.2 MB long, the manager would take about 25 minutes to deliver the installer. This explains the timeout.

It seems that this environment provides a very slow network throughput. The manager sends the WPK file using a stop-and-wait schema, which might be slow (while it guarantees no flooding).

## Fix

For this particular case, I suggest applying this configuration:
```xml
<agent-upgrade>
  <chunk_size>16384</chunk_size>
</agent-upgrade>
```

## Test

```sh
# /var/ossec/bin/agent_upgrade -a 001 -f /home/ec2-user/wazuh_agent_v4.3.0_macos_x86_64.wpk -x upgrade.sh -F

Upgrading...

Upgraded agents:
        Agent 001 upgraded: Wazuh v4.3.0 -> Wazuh v4.3.0
```
